### PR TITLE
[SERV-1114] Rework services, including auth cookie and token services

### PR DESCRIPTION
**Pull request by ksclarke**
* Reorganize the services class hierarchy
* Add additional auth cookie fields
* Change deserialization to use profiles/context instead of type
* Add default type to service interface and overrides when needed
* Change GeoJSONService to GeoJsonService for naming consistency
* Rework "other" services to use an abstract class
* Add explicit serialization for profiles
* Add convenience methods for profile selection
* Clean up some services-related tests

### CODE REVIEW

File: src/main/java/info/freelibrary/iiif/presentation/v3/Service.java [hunk 1]
```diff
@@ -38,9 +43,35 @@
...
+        return getClass().getSimpleName();
+    }
+
+    /**
+     * Sets the service type.
+     *
+     * @param aType A service type
+     * @return The service
+     */
+    T setType(String aType);
```
Comment (markmatney) [inline src/main/java/info/freelibrary/iiif/presentation/v3/Service.java:59]: I feel like `setType` shouldn't be in the Service interface since the majority of the implementing classes make it a no-op (and since AuthTokenService1 doesn't have a type property [per the spec](https://iiif.io/api/auth/1.0/#service-description-1)). Could it instead be only defined on those implementations that _do_ use it (which by my count is GeoJsonService, OtherService2, and OtherService3)?

File: src/main/java/info/freelibrary/iiif/presentation/v3/ServiceDeserializer.java [hunk 2]
```diff
@@ -82,35 +83,42 @@
...
      */
-    private Service<?> deserializeServiceNode(final JsonParser aParser, final JsonNode aNode)
+    @SuppressWarnings(PMD.CYCLOMATIC_COMPLEXITY)
+    private Service<?> deserializeServiceNode(final JsonParser aParser, final JsonNode aNode) // NOPMD
             throws JsonProcessingException {
         final Service<?> service;
 
         if (aNode.isTextual()) {
-            service = new OtherService3(aNode.textValue(), OtherService.class.getSimpleName());
+            service = new OtherService3(aNode.textValue(), null);
```
Comment (markmatney) [inline src/main/java/info/freelibrary/iiif/presentation/v3/ServiceDeserializer.java:92]: If the type property can be null, should `getType` return an `Optional` (like `AbstractOtherService.getProfile`) to protect users from NPE?

File: src/main/java/info/freelibrary/iiif/presentation/v3/ServiceDeserializer.java [hunk 3]
```diff
@@ -120,6 +128,24 @@
...
+     * @param aNode A JSON node
+     * @param aID A service ID
+     * @return The GeoJSON service
+     */
+    private Service<?> deserializeGeoJsonService(final JsonNode aNode, final URI aID) {
+        final GeoJsonService service = new GeoJsonService(aID);
+        final JsonNode typeNode = aNode.get(JsonKeys.TYPE);
+
+        if (typeNode != null) {
+            service.setType(typeNode.asText(null));
```
Comment (markmatney) [inline src/main/java/info/freelibrary/iiif/presentation/v3/ServiceDeserializer.java:143]: Same concern here as on [L92](https://github.com/UCLALibrary/jiiify-presentation/pull/57/files#diff-78aa2c283600bf0eb8b45d655ef231cd69844d2d582b23b18284392b338a131dR92)

File: src/test/java/info/freelibrary/iiif/presentation/v3/services/auth/AuthTokenService1Test.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+import info.freelibrary.iiif.presentation.v3.utils.JSON;
+
+/**
+ * Tests the authorization token service class.
+ */
+public class AuthTokenService1Test {
+
+    private static final String JSON_PATTERN = "{\"@id\":\"{}\",{},\"profile\":\"http://iiif.io/api/auth/1/token\"}";
+
+    private static final String SERVICE_TYPE = "\"@type\":\"" + AuthTokenService1.class.getSimpleName() + "\"";
```
Comment (markmatney) [inline src/test/java/info/freelibrary/iiif/presentation/v3/services/auth/AuthTokenService1Test.java:28]: The spec doesn't say anything about `@type` on access token services. Why is this here?

File: src/main/java/info/freelibrary/iiif/presentation/v3/Service.java (hunk 1 above)
Comment (ksclarke) [reply src/main/java/info/freelibrary/iiif/presentation/v3/Service.java:59]: I removed type but then added it back. I'm still not sure I have it right (I personally think there are some gaps between the specs still), but the spec does say:

`Each object will have properties depending on the service’s definition, but must have either the id or @id and type or @type properties.`

For the older services that were written before the v3 spec:

`For cross-version consistency, this specification defines the following values for the type or @type property for backwards compatibility with other IIIF APIs. Future versions of these APIs will define their own types. These type values are necessary extensions for compatibility of the older versions.`

So while the AuthTokenService1 spec doesn't mention type, the v3 presentation spec is shoehorning it in to conform to its rules.

Comment (ksclarke) [reply src/main/java/info/freelibrary/iiif/presentation/v3/Service.java:59]: I brought this up in the IIIF Slack channel and still need to submit a ticket to them pointing out some of the services specs / docs that are out of step with each other.

File: src/main/java/info/freelibrary/iiif/presentation/v3/ServiceDeserializer.java (hunk 2 above)
Comment (ksclarke) [reply src/main/java/info/freelibrary/iiif/presentation/v3/ServiceDeserializer.java:92]: Good point. I think given that the v3 spec says that the type is required that I should go back to OtherService type when the user doesn't supply a type.

File: src/test/java/info/freelibrary/iiif/presentation/v3/services/auth/AuthTokenService1Test.java (hunk 4 above)
Comment (ksclarke) [reply src/test/java/info/freelibrary/iiif/presentation/v3/services/auth/AuthTokenService1Test.java:28]: See above comment about shoehorning types for older specs into v3.

File: src/main/java/info/freelibrary/iiif/presentation/v3/services/auth/AuthCookieService1.java [hunk 5]
```diff
@@ -49,33 +77,23 @@ public AuthCookieService1(final URI aID) {
...
      *
      * @param aProfile A service profile
-     * @param aID A service ID
-     * @param aLabel A service label
-     * @param aServiceArray An array of related services
+     * @param aID A service ID in string form
      */
-    public AuthCookieService1(final AuthCookieService1.Profile aProfile, final URI aID, final String aLabel,
-            final Service<?>... aServiceArray) {
+    public AuthCookieService1(final AuthCookieService1.Profile aProfile, final URI aID) {
```
[the remark this answers is not in the document]
Comment (ksclarke) [reply src/main/java/info/freelibrary/iiif/presentation/v3/services/auth/AuthCookieService1.java:95]: I have mixed opinions on this but, yes, that's what we said so I'll add a constructor with those two. Thanks.

File: src/test/java/info/freelibrary/iiif/presentation/v3/services/auth/AuthTokenService1Test.java (hunk 4 above)
Comment (markmatney) [reply src/test/java/info/freelibrary/iiif/presentation/v3/services/auth/AuthTokenService1Test.java:28]: Okay, having a type property here makes sense now. Thanks for pointing that out.

File: src/main/java/info/freelibrary/iiif/presentation/v3/ServiceDeserializer.java (hunk 3 above)
Comment (ksclarke) [reply src/main/java/info/freelibrary/iiif/presentation/v3/ServiceDeserializer.java:143]: Implementing the null check at getType() rather than when we set a value here.

File: src/main/java/info/freelibrary/iiif/presentation/v3/ServiceDeserializer.java (hunk 2 above)
Comment (ksclarke) [reply src/main/java/info/freelibrary/iiif/presentation/v3/ServiceDeserializer.java:92]: Implementing the null check at getType() rather than when we set a value here.

**Review by markmatney — APPROVED** [review]

**Review by DRickard — APPROVED** [review]
